### ISSUE TITLE
Is weight decay implemented correctly?

### ISSUE THREAD

**Issue by amujika** [torch/optim#101]
As far as I understand the point of weight decay is to avoid weights being too big (in absolute value).
Shouldn't weight decay be implemented with absolute values to avoid negative values getting larger in magnitude? Otherwise, if we have a large negative weight the gradient will make it try to become even larger.

Current implementation:
`dfdx:add(wd, x)`

How it should be:
`dfdx:add(wd, torch.abs(x))`

This applies to the Adagrad and SGD weight decay.

**Comment by jonathantompson** [comment]
Weight decay is typically L2 norm not L1.  The derivative of L2 is signed,
hence the lack of abs.

**Comment by amujika** [comment]
You're right, I got confused. Thanks for the response!

**Comment by iammarvelous** [comment]
@jonathantompson Hi, I got confused here. Is weight decay the hyperparameter of L2 regularization?
 If so, it does not make sense using `dfdx:add(wd, x)`. Instead, shouldn't we use `dfdx:add(wd*x)`?

**Comment by soumith** [comment]
dfdx:add(wd, x) will do: dfdx += wd * x. It's one of the function signatures of the add function, when wd is scalar.
